Don't use _ASSERTE in runtime tests

lambdageek opened the pull request:
In debug configurations, use of `_ASSERTE` requires the `PAL_frpintf` function and linking with the `coreclrpal` static library.

Instead just use `assert()` from `<assert.h>`

Fixes https://github.com/dotnet/runtime/issues/68279

lambdageek committed:
link coreclrpal static library into tests that use pal.h

In debug configurations, use of assert() (aka _ASSERTE) requires the
PAL_frpintf function

Fixes https://github.com/dotnet/runtime/issues/68279

lambdageek commented on src/tests/CMakeLists.txt:
```diff
@@ -14,6 +14,16 @@ include_directories("${CLR_SRC_NATIVE_DIR}")
 add_subdirectory(../coreclr/pal/prebuilt/inc ../coreclr/pal/prebuilt/inc)
 
+
+# tests in the Debug configuration depend on the coreclrpal static library if they use assert
+if (NOT CLR_CMAKE_HOST_WIN32)
+  # Add component names that the pal build expects
+  add_component(${CMAKE_INSTALL_DEFAULT_COMPONENT_NAME})
+  add_definitions(-DUSE_STL)
+  include_directories(../coreclr/pal/inc/rt ../coreclr/pal/inc ../coreclr/inc)
+  add_subdirectory(../coreclr/pal/src ../coreclr/pal/src)
+endif()
```
I don't know if this is the proper way to get the coreclrpal target to build.  It works but I'm not sure if i'm doing something wrong here.

janvorli replied:
I am a bit confused. If the profiler test needs anything from PAL, then it would not build before. Let me try to understand what's going on.

janvorli replied:
Ah, I see, it is using some headers that pull in the PAL headers and with the change in #64631, the _ASSERTE references the two PAL functions. I don't think we should let the profiler test pull in the PAL just because of that. I would replace the usage of _ASSERTE by assert in the test instead.

lambdageek commented:
Do we have a CI leg that will build the runtime tests in the debug configuration?

trylek reviewed (approved):
LGTM but I'd welcome confirmation from JanV who's the PAL expert.

trylek commented:
AFAIK we only do checked configurations in the lab as debug is prohibitively slow.

am11 commented on src/tests/CMakeLists.txt:
```diff
@@ -14,6 +14,16 @@ include_directories("${CLR_SRC_NATIVE_DIR}")
+  # Add component names that the pal build expects
+  add_component(${CMAKE_INSTALL_DEFAULT_COMPONENT_NAME})
+  add_definitions(-DUSE_STL)
+  include_directories(../coreclr/pal/inc/rt ../coreclr/pal/inc ../coreclr/inc)
+  add_subdirectory(../coreclr/pal/src ../coreclr/pal/src)
```
We try to use absolute paths as much as possible. Could you please find and replace them in this file:

```sh
if [ "$(uname)" = "Darwin" ]; then space=' '; fi
sed -i$space'' 's|../coreclr|${CLR_REPO_ROOT_DIR}/src/coreclr|g' src/tests/CMakeLists.txt
```

lambdageek committed:
don't link coreclrpal static lib into the profiler tests

lambdageek committed:
Don't use _ASSERTE in runtime tests

That macro requires linking the coreclrpal static library, which the
runtime tests don't currently do.

lambdageek commented:
I'm going to try to change this PR so that the profiler tests don't depend on the PAL.

lambdageek commented:
> AFAIK we only do checked configurations in the lab as debug is prohibitively slow.

Maybe we should do a debug run with Mono - we don't have as noticeable a difference, I think (nearly all of our asserts are on in Release, too). /cc @steveisok - maybe a Linux x64 debug build that runs the runtime tests (also debug) in the rolling builds?

lambdageek committed:
second arg of add_subdirectory is a binary dir

lambdageek commented on src/tests/CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ include(${CLR_ENG_NATIVE_DIR}/configurecompiler.cmake)
 
 # Add this subdir. We install the headers for the jit.
-add_subdirectory(../coreclr/pal/prebuilt/inc ../coreclr/pal/prebuilt/inc)
+add_subdirectory(${CLR_REPO_ROOT_DIR}/src/coreclr/pal/prebuilt/inc ../coreclr/pal/prebuilt/inc)
```
Leaving out the binary dir is an error:
```
CMake Error at CMakeLists.txt:15 (add_subdirectory):
  add_subdirectory not given a binary directory but the given source
  directory
  "/home/aleksey/working/dotnet-runtime/src/coreclr/pal/prebuilt/inc" is not
  a subdirectory of "/home/aleksey/working/dotnet-runtime/src/tests".  When
  specifying an out-of-tree source a binary directory must be explicitly
  specified.
```

Using `${CLR_ARTIFACTS_OBJ_DIR}/coreclr/pal/prebuilt/inc` works, except I'm not actually convinced it's the right thing.  the normal coreclr cmake stuff goes to `artifacts/obj/coreclr/<<Linux.x64.Debug>>/pal/prebuilt/inc`.  so we're making another copy.  In which case it actually feels more correct to just drop it into a subdirectory of the runtime tests' binary dir.  somehting like `add_subdirectory (${CLR_REPO_ROOT_DIR}/src/coreclr/pal/prebuilt/inc coreclrpal)`

lambdageek committed:
use an absolute path for binary_dir
